validate-index.cpp

```diff
@@ -0,0 +1,92 @@
+#include "clang/Index/IndexDataStore.h"
+#include "clang/Index/IndexUnitReader.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace llvm;
+using namespace llvm::sys;
+using namespace clang::index;
+
+static cl::opt<std::string> IndexStore(cl::Positional, cl::Required,
+                                       cl::desc("<indexstore>"));
+
+int main(int argc, char **argv) {
+  cl::ParseCommandLineOptions(argc, argv);
+
+  std::string storeError{};
+  auto store = IndexDataStore::create(IndexStore, storeError);
+  if (not store) {
+    errs() << "error: failed to open indexstore " << IndexStore << " -- "
+           << storeError << "\n";
+    return EXIT_FAILURE;
+  }
+
+  std::vector<std::string> unitNames{};
+  store->foreachUnitName(false, [&](StringRef unitName) {
+    unitNames.push_back(unitName.str());
+    return true;
+  });
+
+  auto exitStatus = EXIT_SUCCESS;
+  for (const auto &unitName : unitNames) {
+    std::string readerError;
+    auto reader = IndexUnitReader::createWithUnitFilename(unitName, IndexStore,
+                                                          readerError);
+    if (not reader) {
+      errs() << "error: failed to read unit file " << unitName << " -- "
+             << readerError << "\n";
+      return EXIT_FAILURE;
+    }
+
+    const std::vector<std::pair<std::string, std::string>> unitPaths = {
+        {"MainFilePath", reader->getMainFilePath()},
+        {"SysrootPath", reader->getSysrootPath()},
+        {"WorkingDirectory", reader->getWorkingDirectory()},
+        // TODO: OutputFile does not need to exist, but its path needs to match
```

Comment:
might be good to have flags to enable/disable all/some of these checks

Reply:
My thinking was: print all checks and let user filter with grep. In my test use, I was always filtering out `OutputFile`, which is why I commented this out. It's possibly worth doing, I guess the question is grepping fine now, or does this need flags now? For my needs, I grepping does the job.

Are you thinking a flag for each kind of file? Would they all be on be default, and then the flags would opt out? Or something else?